content/post/2017-09-14-using-ecto-and-sqlite3-with-nerves.md

```diff
@@ -0,0 +1,202 @@
+we unfortunately don't have the luxury of Mix Tasks. We are going to have to do something a little custom.
+
+Note: There is a number of ways to possibly accomplish getting your database setup in Nerves,
+and this is by no means the only way.
+
+Naturally we should just be able to run the Mix Tasks from our application code.
+Unfortunately this won't work. The Ecto tasks rely on things that just aren't available
+in our Nerves release. So we will have to implement them a little bit manually.
+
+First make sure we have a migration.
+`mix ecto.gen.migration add_some_stuff`
+Edit this file accordingly. (leaving it empty is fine too.)
+
+In our `application.ex` file again lets add some functionality.
+
+```elixir
+@otp_app Mix.Project.config[:app]
+def start(_type, _args) do
+  import Supervisor.Spec, warn: false
+  :ok = setup_db!()
+  children = [
+    supervisor(HelloDb.Repo, [])
+  ]
+
+  opts = [strategy: :one_for_one, name: HelloDb.Supervisor]
+  Supervisor.start_link(children, opts)
+end
+
+defp setup_db! do
+  repos = Application.get_env(@otp_app, :ecto_repos)
+  for repo <- repos do
+    setup_repo!(repo)
+    migrate_repo!(repo)
+  end
+  :ok
+end
+
+defp setup_repo!(repo) do
+  db_file = Application.get_env(@otp_app, repo)[:database]
+  unless File.exists?(db_file) do
+    :ok = repo.__adapter__.storage_up(repo.config)
+  end
+end
+
+defp migrate_repo!(repo) do
+  opts = [all: true]
+  {:ok, pid, apps} = Mix.Ecto.ensure_started(repo, opts)
+
+  migrator = &Ecto.Migrator.run/4
+  pool = repo.config[:pool]
+  migrations_path = Path.join((:code.priv_dir(@otp_app) |> to_string), "repo")
+  migrated =
+    if function_exported?(pool, :unboxed_run, 2) do
+      pool.unboxed_run(repo, fn -> migrator.(repo, migrations_path, :up, opts) end)
+    else
+      migrator.(repo, migrations_path, :up, opts)
+    end
+
+  pid && repo.stop(pid)
+  Mix.Ecto.restart_apps_if_migrated(apps, migrated)
+end
+```
+
+
+We can brake that down a bit here:
+
+The `setup_repo!/1` was derived from the (create)[https://github.com/elixir-ecto/ecto/blob/master/lib/mix/tasks/ecto.create.ex]
+mix task. It just checks for the database file's existance, and creates it if the file
+is non existant.
+
+The `migrate_repo/1` function is a bit more interesting. We actually need to start
+the repo (and its pool), find the path to our migrations, then of course run the migrations,
+and finally restart everything. Lucily `Mix.Ecto` is available for us that does
+much of the hard work for us.
+
+And there we have it, Your Sqlite repo will be setup and migrated at application startup.
+Obviously this is a little bit more config than your average Elixir or even Phoenix
+setup, but It's all fairly straight forward. Hopefully this gives a good
+jumping off point to storing data on a Nerves project.
+
+
+## Bonus points
+Obviously some things weren't concidered here. Be sure to concider some things
```

Comment:
~~concidered~~ considered and ~~concider~~ consider

Reply:
Thanks :)